Fix documentation - remove extra periods between paragraphs

chmreid opened the pull request:
This will address one of the items brought up in issue #443 

The documentation has [a paragraph](https://hca.readthedocs.io/en/latest/cli.html#download-manifest) with stray commas everywhere, this fixes the problem

chmreid commented:
@jessebrennan by "the way the help text is displayed" do you mean the help printed on the command line? (Is that what you pasted above?)

This PR addresses an issue with the sphinx documentation. IMHO the `--help` flag should be able to handle two consecutive line breaks. There are also multiple docstrings that are formatted the same as the one I've changed (for example, the docstring of the function right above this one), so is the `--help` flag breaking the output for those too?

jessebrennan commented:
@chmreid, sorry yes I should have been more clear. If you run

```
hca dss download-manifest --help
```

You will see the formatting of the output is messed up by your change.

>This PR addresses an issue with the sphinx documentation. IMHO the --help flag should be able to handle two consecutive line breaks. There are also multiple docstrings that are formatted the same as the one I've changed (for example, the docstring of the function right above this one), so is the --help flag breaking the output for those too?

I don't understand what you are saying here.

chmreid commented:
To clarify what I was asking:

I'm looking in the same file that I modified in this PR, and looking at the function `download()` right above the function `download_manifest()` that I modified, and looking at the docstring of the `download()` function: https://github.com/HumanCellAtlas/dcp-cli/blob/7213d9505a0d9d673bab1901224fdfd7945b3d6a/hca/dss/__init__.py#L202-L233

That docstring is formatted exactly the same way as my new proposed/changed docstring. So I was wondering  askingif using the command line interface with the `--help` flag to display help for `download()` messes up the formatting the same way that the `--help` flag to display help for `download_manifest()` messes up the formatting.

----

To try and answer this question, I ran the `hca dss download --help` command and saw that in fact the `--help` flag is turning any arbitrary number of newline characters into a single newline character, so the docstring

```
        Download a bundle and save it to the local filesystem as a directory.
        By default, all data and metadata files are downloaded. To disable the downloading of data files,
        use `--data-files ''` if using the CLI (or `data_files=()` if invoking `download` programmatically).
        Likewise for metadata files.
        If a retryable exception occurs, we wait a bit and retry again.  The delay increases each time we fail and
        decreases each time we successfully read a block.  We set a quota for the number of failures that goes up with
        every successful block read and down with each failure.
```

is rendered as:

```
Download a bundle and save it to the local filesystem as a directory.
By default, all data and metadata files are downloaded. To disable the downloading of data files,
use --data-files '' if using the CLI (or data_files=() if invoking download programmatically).
Likewise for metadata files.
If a retryable exception occurs, we wait a bit and retry again.  The delay increases each time we fail and
decreases each time we successfully read a block.  We set a quota for the number of failures that goes up with
every successful block read and down with each failure.
```

Note the missing two newlines.

----

If I add a bullet list to the docstring, like so:

```
        Download a bundle and save it to the local filesystem as a directory.

        By default, all data and metadata files are downloaded. To disable the downloading of data files,
        use `--data-files ''` if using the CLI (or `data_files=()` if invoking `download` programmatically).
        Likewise for metadata files.

        - I will not talk during class
        - I will not talk during class
        - I will not talk during class
        - I will not talk during class
        - I will not talk during class

        If a retryable exception occurs, we wait a bit and retry again.  The delay increases each time we fail and
        decreases each time we successfully read a block.  We set a quota for the number of failures that goes up with
        every successful block read and down with each failure.
```

here is the corresponding output:

```
Download a bundle and save it to the local filesystem as a directory.
By default, all data and metadata files are downloaded. To disable the downloading of data files,
use --data-files '' if using the CLI (or data_files=() if invoking download programmatically).
Likewise for metadata files.
I will not talk during class

I will not talk during class

I will not talk during class

I will not talk during class

I will not talk during class
If a retryable exception occurs, we wait a bit and retry again.  The delay increases each time we fail and
decreases each time we successfully read a block.  We set a quota for the number of failures that goes up with
every successful block read and down with each failure.
```

So, the docstrings are being formatted using a kind of quasi-ReST (restructured text) so that Sphinx can interpret it and turn it into nice fancy HTML. But the command line `--help` option does not do that, and mangles it in some way that does not preserve formatting.

Will keep looking into this to see if there is a solution that works for both. Ultimately whatever the `--help` flag does to mangle docstrings before displaying them should be fixed, but it's uncertain how easy/difficult that will be.

chmreid commented:
Looking into this more, in [`util/_docs.py`](https://github.com/HumanCellAtlas/dcp-cli/blob/master/hca/util/_docs.py) there is a utility `_parse_docstring()`, which uses Sphinx's ReST parser to parse the docstrings, and that function is used to create help for the command line utilities in [`hca/util/__init__.py`](https://github.com/HumanCellAtlas/dcp-cli/blob/master/hca/util/__init__.py#L606).

This `_parse_docstring()` function is where the ReST bullet list is converted to plain text (and mangled in the process). From this debugger session you can see that the RSTParser object identifies this portion of the docstring as a bullet list:

```
(Pdb) where
  /Users/charles/codes/dcp-cli/vp/bin/hca(11)<module>()
-> load_entry_point('hca==6.4.0', 'console_scripts', 'hca')()
  /Users/charles/codes/dcp-cli/vp/lib/python3.6/site-packages/hca-6.4.0-py3.6.egg/hca/cli.py(108)main()
-> parser = get_parser(help_menu=True)
  /Users/charles/codes/dcp-cli/vp/lib/python3.6/site-packages/hca-6.4.0-py3.6.egg/hca/cli.py(97)get_parser()
-> dss_cli.add_commands(parser._subparsers, help_menu=help_menu)
  /Users/charles/codes/dcp-cli/vp/lib/python3.6/site-packages/hca-6.4.0-py3.6.egg/hca/dss/cli.py(12)add_commands()
-> dss_cli_client.build_argparse_subparsers(dss_subparsers, help_menu=help_menu)
  /Users/charles/codes/dcp-cli/vp/lib/python3.6/site-packages/hca-6.4.0-py3.6.egg/hca/util/__init__.py(612)build_argparse_subparsers()
-> method_args = _parse_docstring(docstring)
> /Users/charles/codes/dcp-cli/vp/lib/python3.6/site-packages/hca-6.4.0-py3.6.egg/hca/util/_docs.py(92)_parse_docstring()
-> if node.tagname is 'paragraph' and method_args['summary'] == '':

(Pdb) p node
<bullet_list: <list_item...><list_item...><list_item...><list_item...> ...>
```

But the `_parse_docstring()` function does not treat bullet lists differently, so it ends up converting each bullet as separate paragraphs, with no leading/trailing newline, as mentioned above:

```
(Pdb) p node.astext()
'bundle_uuid - the UUID of the bundle containing the file in DSS.\n\nbundle_version - the version of the bundle containing the file in DSS.\n\nfile_name - the name of the file as specified in the bundle.\n\nfile_uuid - the UUID of the file in the DSS.\n\nfile_sha256 - the SHA-256 hash of the file.\n\nfile_size - the size of the file.'
```

To resolve this, I wrote a function that lives next to `_parse_docstring()` that turns a bullet list into text and added it to this PR.

chmreid commented:
Rebasing onto master
